### ISSUE TITLE
indigo-cool-vivid-60 not passing contrast guidelines

### ISSUE THREAD

**Issue by darekkay** [uswds/uswds#3329]
The [documentation](https://designsystem.digital.gov/design-tokens/color/overview/) states:

> A magic number of 50+ results in WCAG 2.0 AA contrast or AAA Large Text contrast (example: gray-90 and red-40).

I've checked all color combinations and `indigo-cool-vivid-60` (`#4150f2`) does not always meet this criteria. Invalid combinations:

- `gray-cool-10` (`#dcdee0`): [4.26](https://contrast-ratio.com/#%23dcdee0-on-%234150f2)
- `green-vivid-10` (`#c3ee90`): [4.36](https://contrast-ratio.com/#%23c3ee90-on-%234150f2)
- `mint-vivid-10` (`#5dfdc8`): [4.47](https://contrast-ratio.com/#%235dfdc8-on-%234150f2)

I think a luminance/lightness bounds table per grade would be great (I couldn't find one here and I've tried to reach out to @thisisdano regarding this). 

Related to: #3094

**Comment by thisisdano** [comment]
Thanks for pointing this out. I'll check our own checker to see why we weren't able to catch this in our own tests, then follow up.

**Comment by thisisdano** [comment]
So, I'm not yet sure why these combinations didn't trigger our tester, but they did not. And you're right, they are invalid combinations.

After doing the initial color work, I developed the following table with luminance bounds. The `range` column in this table shows the bounds (x100). Does this guidance square with your findings? If so, I will go back and check our values against this guidance, and make adjustments.

<img width="978" alt="Screen Shot 2020-02-28 at 10 00 55 AM" src="https://user-images.githubusercontent.com/11464021/75573381-92ad5b80-5a11-11ea-87e0-22ce230cb372.png">

I do see that `indigo-cool-vivid-60` (`#4150f2`), `green-vivid-10` (`#c3ee90`), and `gray-cool-10` (`#dcdee0`) all fall outside the guidance bounds. I'm not sure why the checker didn't catch these combinations, so that'll be the next thing we investigate. 

Once again, thanks for bringing this to my attention — I'm glad we're able to improve this system.

**Comment by darekkay** [comment]
Thanks @thisisdano for your answer and the overall commitment in this project!

I learned about the USWDS color palette from @pspeter3. He came up with his own luminance bounds in his [blog post](https://pspeter3.com/blog/2020/02/19/accessible-contrast-shades/). This table overlaps with yours, but there are some major differences. For example, the luminance for violet-warm-40 (`#bf77c8`) is `0.2844`, which doesn't fit into his table.

Do you remember how you came up with your bounds? Did you use [this formula](https://github.com/One-com/one-color/blob/master/lib/plugins/luminance.js) to calculate the luminance? I would like to invest some more time in this area and document the results.

**Comment by darekkay** [comment]
I am using the following formula to calculate the contrast ratio:

```javascript
if (lum1 > lum2) {
  return (lum1 + 0.05) / (lum2 + 0.05);
}

return (lum2 + 0.05) / (lum1 + 0.05);
```

I've created [a script](https://gist.github.com/darekkay/4880a4d57734ca58a021124317235ec2) to calculate the min/max bounds per grade, and here are my results:

```
  "0":  [1, 1],
  "10": [0.999, 0.6499],
  "20": [0.65, 0.6249],
  "30": [0.625, 0.3],
  "40": [0.3001, 0.25],
  "50": [0.1833, 0.175],
  "60": [0.175, 0.1],
  "70": [0.0999, 0.05],
  "80": [0.0499, 0.0465],
  "90": [0.0464, 0.001],
  "100": [0, 0]
```

There are values that are equal for all 3 tables, but also some major differences. Let's inspect one value that is different.

Grade 50 is equal across all 3 tables. Let's calculate the minimum luminance for grade 10
  - Max luminance for grade 50 = 0.183 (equal for all tables)
  - Contrast ratio between 50 and 10 >= 3

Min luminance for grade 10:

```
  3 = (x + 0.05) / (0.183 + 0.05)
  x = 0.649
```

However, this is the value you've got for grade 20.

I've checked your table as well and it is accessible in itself for all the bounds. I think, it's just more restrictive, as it has more gaps between two grades (e.g. how would you place a color with a luminance of 0.46? it's between 20 and 30).

Is there any mistake in my calculation/approach? How did you derive your values?

**Comment by mejiaj** [comment]
@darekkay 
CC @thisisdano 

There's a [check-contrast](https://github.com/uswds/uswds-site/wiki/Contrast-checker-instructions) tool that's used to pass the tokens and check for several things. You can find that tool [here](https://github.com/uswds/uswds-site/blob/master/utils/check-contrast.js).

Getting the luminance of `indigo-cool-vivid` family for example would look like:
```bash
npm run check-contrast -- --l --f indigo-cool-vivid
```

As you can see in the code for `check-contrast.js` it's using [chroma-js](https://gka.github.io/chroma.js/) to convert colors to RGB and get their luminance via chroma's luminance function.

Here's the function that gets run to get luminance:
```js
const luminanceForFamily = (family) => {
  return family.colors.map((color) => {
    return {
      colorName: formatColorName(family.name, color.grade),
      relativeLum: chroma(color.value).luminance(),
    }
  });
};

```

Which results in something like:
```js
[
  {
    colorName: 'indigo-cool-vivid-40',
    relativeLum: 0.28474325989959903
  },
  {
    colorName: 'indigo-cool-vivid-60',
    relativeLum: 0.13271943815711382
  }
]
```

Chroma seems to be using similar values for the formula. Do you believe @darekkay we'd need to change the way we're using chroma to get the same values as you?

TL;DR we need to figure out if we're using chroma correctly and what changes are needed.

We also looking into the contrast issue and why it didn't catch it.

**Comment by thisisdano** [comment]
I believe the luminance values we're getting from chroma are fine and are consistent with what @darekkay is seeing. An open question is why the tester didn't catch the contrast fail between [failing pairs](https://contrast-ratio.com/#%23dcdee0-on-%234150f2) like `indigo-cool-vivid-60` and `gray-cool-10`.

**Comment by thisisdano** [comment]
A second issue is around our luminance range guidance for color grades: why is it what it is and how well our current values square with that guidance. This is a bit of a larger question. So, graphs! I've tried to look at our current draft guidance for grades and luminance ranges against the @darekkay's proposal (I know it's not really an official proposal, but bear with me).

I wanted to visualize how they differed, and start to work out our reasoning.

And before I do all this, I want to say that I love the work @darekkay has done here, and I appreciate the opportunity to think this through in public.

The big differences here are that the proposal optimizes for coverage (i.e. increasing how much of the spectrum is covered in the ranges) and USWDS optimizes for consistency (i.e. how similar colors are within a range and how similar ranges are between each other). The proposal covers more of the spectrum, USWDS is more consistent.

I've called out three instances in the proposal where the ranges are particularly unequal, either unusually small (grades `20`, and `80`) or unusually large (grade `30`). I'm not qualified to speak to the algorithm that generated the proposal, but I wonder how we could modify it to counter these anomalies. 

#### USWDS and proposed grades and ranges
![range-charts-proposal](https://user-images.githubusercontent.com/11464021/75909542-219ae900-5e01-11ea-9309-fb1bcc5f8178.png)

It's important that our color system be consistent and predictable, that users know what they're getting when they choose a color of a certain grade. This makes us inclined to favor smaller, more equal ranges, and consistent spacing between ranges. I'm showing the difference between the two methods in a couple ways: graphed against luminosity and against lightness (defined as the L in the HSL model), and I've limited it to the gray family so we don't have to worry about hue.

Luminosity is a geometric function and lightness is a linear function so it's easier (in a linear scale) to see the evenness of a color system mapped to lightness. (An increment of luminance is equal to a smaller increment of lightness at the light end of the spectrum than at the dark end. Is this because our eyes are more sensitive to light colors?)

#### Luminance comparison
![proposal-comparison-luminance](https://user-images.githubusercontent.com/11464021/75909548-252e7000-5e01-11ea-9c86-c6c9140690e4.png)

#### Lightness comparison
![proposal-comparison-lightness](https://user-images.githubusercontent.com/11464021/75909552-26f83380-5e01-11ea-8a32-654e5ab5a1fc.png)

So while the USWDS model has more gaps (less of the spectrum is covered by USWDS grades), I would say that it is better suited to the goals of the design system: by providing a clearer, more obvious relationship between grades; and by assuring that colors of a certain grade are fairly consistent in weight between color family (that is, assuring that `indigo-30` isn't significantly more luminous that `magenta-30`.

USWDS is structured around a relatively static collection of design tokens. We expect our users to use these tokens, and we don't expect the values of our tokens to change very frequently. Our users expect us to be predictable, if we _do_ change the value of a specific token, they would expect that this change would not have a dramatic effect of the tone of their site. One of the very nice things about color tokens is that the system can adapt and update the values under the hood but these changes will never adversely affect either accessibility or broad tonal intent (that is, blue will still be "blue" and `blue-40` will always have a predictable luminosity).

I think a different system with different design principles might benefit from rubric that optimizes for spectrum coverage, minimizing the times a user would get a `color does not match any grade` error!

- - -

**I should note that the USWDS range guidance I've outlined above is still a work in progress.** There may be some errors! We do expect to publish it this month, but the values may change. We'll be using this guidance to build out the vivid palette ([issue](https://github.com/uswds/uswds/issues/3094)). Also, there are a few current token values that fall outside this guidance, and we'll be bringing them into conformance as well. Currently:

- `gray-cool-10`
- `green-10v`
- `green-30v`
- `cyan-20`
- `cyan-30`
- ...and of course `indigo-cool-60v`

**Comment by mejiaj** [comment]
Taking the example of `indigo-cool-vivid-60` and `gray-cool-10`.

If we use `check-contrast.js` in uswds-site.
```bash
npm run check-contrast -- --c --f indigo-cool-vivid
``` 

For indigo-cool-vivid-40 tested on white. We have this contrastResult:
```js
ContrastResult {
  ratio: 3.1367323133404716,
  base: 'indigo-cool-vivid-40',
  contrast: 'white-0'
}

if (!contrastResult.isCompliant()) {
    output.push(contrastResult);
}
```

`!contrastResult.isCompliant()` is false on `console.log()`. But the code inside never runs.

We should expect for it to check and fail aaCompliance.

---

Right now the check is only running for `aa` on differences over `50`. In the example above, it would check `aaLarge` and pass just like [here](https://contrast-ratio.com/#%23dcdee0-on-%234150f2).

So we would need to:
1. Figure out why it's not pushing compliance errors
1. Define criteria for our tests 

I imagine if we set a rule like fail everything if `aa` fails would result in quite a bit of errors when running
```bash
// check contrast grades between all families
npm run check-contrast -- --ca --d
```

**Comment by pspeter3** [comment]
@thisisdano thank you for the great explanation and charts! @darekkay thank you for opening this issue and linking to my post!

Is there any additional information I can provide?

**Comment by darekkay** [comment]
Thanks for all the information you've provided, that's really helpful!

**Comment by pspeter3** [comment]
Have you considered writing a post explaining your table?

**Comment by thisisdano** [comment]
I would love to write a post about this — I've found my writing time limited recently but perhaps this issue has gotten me most of the way there!

**Comment by pspeter3** [comment]
Awesome! I would be happy if I can!

When I generated my table, I went for widest possible range but I agree that it doesn't necessarily help developers think about consistency. Your tighter ranges help with that a lot.

**Comment by darekkay** [comment]
I don't know as much about color theory as Dan, but I've summarized everything I've learned about the "magic numbers" concept in a [blog post](https://darekkay.com/blog/accessible-color-palette/) 🙂
